lib/cli/src/wasmer_env.rs

```diff
@@ -0,0 +1,127 @@
+use wasmer_registry::WasmerConfig;
+
+/// Command-line flags for determining the local "Wasmer Environment".
+///
+/// This is where you access `$WASMER_DIR`, the `$WASMER_DIR/wasmer.toml` config
+/// file, and specify the current registry.
+#[derive(Debug, Clone, PartialEq, clap::Parser)]
+pub struct WasmerEnv {
+    /// Set Wasmer's home directory
+    #[clap(long, env = "WASMER_DIR", default_value = WASMER_DIR.as_os_str())]
+    wasmer_dir: PathBuf,
+    /// The registry to fetch packages from (inferred from the environment by
+    /// default)
+    #[clap(long, env = "WASMER_REGISTRY")]
+    registry: Option<Registry>,
+    /// The API token to use when communicating with the registry (inferred from
+    /// the environment by default)
+    #[clap(long, env = "WASMER_TOKEN")]
+    token: Option<String>,
+}
+
+impl WasmerEnv {
+    pub(crate) fn new(
+        wasmer_dir: PathBuf,
+        registry: Option<Registry>,
+        token: Option<String>,
+    ) -> Self {
+        WasmerEnv {
+            wasmer_dir,
+            registry,
+            token,
+        }
+    }
+
+    /// Get the GraphQL endpoint used to query the registry.
+    pub fn registry_endpoint(&self) -> Result<Url, Error> {
+        if let Some(registry) = &self.registry {
+            return registry.graphql_endpoint();
+        }
+
+        let config = self.config()?;
+        let url = config.registry.get_current_registry().parse()?;
+
+        Ok(url)
+    }
+
+    /// Load the current Wasmer config.
+    pub fn config(&self) -> Result<WasmerConfig, Error> {
+        WasmerConfig::from_file(self.dir())
+            .map_err(Error::msg)
+            .with_context(|| {
+                format!(
+                    "Unable to load the config from the \"{}\" directory",
+                    self.dir().display()
+                )
+            })
+    }
+
+    /// The directory all Wasmer artifacts are stored in.
+    pub fn dir(&self) -> &Path {
+        &self.wasmer_dir
+    }
+
+    /// The API token for the active registry.
+    pub fn token(&self) -> Option<String> {
+        let config = self.config().ok()?;
+        let login = config.registry.current_login()?;
+        Some(login.token.clone())
+    }
+}
+
+impl Default for WasmerEnv {
+    fn default() -> Self {
+        Self {
+            wasmer_dir: WASMER_DIR.clone(),
+            registry: None,
+            token: None,
+        }
+    }
+}
+
+/// The default value for `$WASMER_DIR`.
+pub static WASMER_DIR: Lazy<PathBuf> =
+    Lazy::new(|| match wasmer_registry::WasmerConfig::get_wasmer_dir() {
+        Ok(path) => path,
+        Err(e) => {
+            if let Some(install_prefix) = std::env::var_os("WASMER_INSTALL_PREFIX") {
+                PathBuf::from(install_prefix)
+            } else {
+                panic!("Unable to determine the wasmer dir: {e}");
+            }
+        }
+    });
+
+/// A registry as specified by the user.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Registry(String);
```

Comment:
Can we make this a `url::Url`?

Reply:
We want users to be able to do `--registry wasmer.wtf` and have `registry.graphql_endpoint()` return `https://registry.wasmer.wtf/graphql`, but `wasmer.wtf` isn't a valid URL.

We could write the `From<String>` impl to automatically call `wasmer_registry::format_graphql()` on whatever string users pass in. However, it'd mean the help text and other printed messages would show `https://registry.wasmer.wtf/graphql` even if the user specified `--registry wasmer.wtf` and that's not great for UX. 

Making `Registry` a newtype wrapper around a `String` is a bit of a compromise that lets us capture the user's original input while still getting the coerced GraphQL endpoint automatically.

Reply:
Yeah, putting it into the `FromStr` impl would have been my idea, and using the magic conversion function. 

The struct could also hold both the original value as string, and the derived value as a url, but that's also not necessarily ideal.

Fine for me as is.